## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

C:\Users\mzkan\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (6,11,12) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
print(contrib.shape)

(1125659, 18)


- **1a answer here:** Yes, it matches expectations, since the data has 1,125,659 rows and 18 columns

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
print(contrib.columns)

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')


- **1b answer here:** Yes, all columns are included

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
contrib.head()

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here:** cand_id match with cand_name, and contbr_st all say CA since it is for California

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [35]:
# 1d YOUR CODE HERE
contrib['election_tp'].unique()

array(['P2016', 'G2016', nan, 'P2020'], dtype=object)

- **1d answer here:** The type of election, for example if its for a primary, general or convention with year. Yes, except one, there are some nan values witch is missing data

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
contrib.dtypes

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- **1e answer here:** contbr_zip is an object, contb_receipt_amt is a float, and contb_receipt_dt is a datetime

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
contrib.count().sort_values(ascending=False)

cmte_id              1125659
contbr_st            1125659
file_num             1125659
form_tp              1125659
cand_id              1125659
cand_nm              1125659
contbr_nm            1125659
contb_receipt_dt     1125659
contb_receipt_amt    1125659
tran_id              1125659
contbr_city          1125633
contbr_zip           1125564
election_tp          1124234
contbr_occupation    1115260
contbr_employer       967757
memo_text             501148
memo_cd               144268
receipt_desc           15045
dtype: int64

- **1f answer here:** cmte_id, contbr_st, file_num, form_tp, cand_id, cand_nm, contbr_nm, contb_receipt_dt, contb_receipt_amt, tran_id. Yes, receipt_desc and memo_cd since more than half the values are nulls

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE
contrib[contrib['cand_nm'] == 'Sanders, Bernard']['cand_id'].unique()

array(['P60007168'], dtype=object)

In [9]:
contrib[contrib['cand_nm'] == 'Sanders, Bernard']['cand_nm'].unique()

array(['Sanders, Bernard'], dtype=object)

- **1g answer here:** No issues, all Bernard Sanders cand_id matches the cand_nm

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [10]:
# 1h YOUR CODE HERE
contrib['contbr_st'].unique()

array(['CA'], dtype=object)

- **1h answer here:** All records are in California

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [11]:
# 1i YOUR CODE HERE
contrib['tran_id'].value_counts()[contrib['tran_id'].value_counts()>1]

A5602AD777C8C4632B5A    4
ADB49CB248C174E298F0    4
SA17.1136499            3
SA17.774344             3
A340DF85B7F884133A20    3
                       ..
SA17A.72592             2
SA17.692303             2
SA17.833690             2
SA17.1072328            2
SA17A.1629012           2
Name: tran_id, Length: 3419, dtype: int64

- **1i answer here:** There are 3,419 duplicate entires. For the SA17 form, there seems to have many duplicates from that form type

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [12]:
# 1j YOUR CODE HERE
len(contrib[contrib['contb_receipt_amt'] < 0])

11896

In [13]:
contrib[contrib['contb_receipt_amt'] < 0].head()

cmte_id    cand_id                    cand_nm           contbr_nm  \
19   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
23   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
81   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
190  C00574624  P60006111  Cruz, Rafael Edward 'Ted'    NOWELL, DIANA L.   
213  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  LICHTY, ANDREW MR.   

                contbr_city contbr_st     contbr_zip  \
19                    CHICO        CA 959,289,507.00   
23                    CHICO        CA 959,289,507.00   
81                    CHICO        CA 959,289,507.00   
190  RANCHO SANTA MARGARITA        CA 926,884,928.00   
213               SAN DIEGO        CA 921,096,720.00   

                        contbr_employer     contbr_occupation  \
19                        SELF EMPLOYED               RANCHER   
23                        SELF EMPLOYED               RANCHER   
81                        SELF EMPLOYED               RANCHER   
190  CAPISTRAND UNIFIED SCHOOL DISTRICT  LIBRARIAN TECHNICIAN   
213                       SELF EMPLOYED           REAL ESTATE   

     contb_receipt_amt contb_receipt_dt              receipt_desc memo_cd  \
19              -25.00       2016-04-29  REDESIGNATION TO GENERAL       X   
23             -150.00       2016-04-29  REDESIGNATION TO GENERAL       X   
81              -60.00       2016-04-14  REDESIGNATION TO GENERAL       X   
190            -100.00       2016-04-11  REDESIGNATION TO GENERAL       X   
213             -25.00       2016-04-30  REDESIGNATION TO GENERAL       X   

                    memo_text form_tp  file_num         tran_id election_tp  
19   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826482B       P2016  
23   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826483B       P2016  
81   REDESIGNATION TO GENERAL   SA17A   1077664   SA17A.1827494       P2016  
190  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1639830B       P2016  
213  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826888B       P2016

- **1j answer here:** There are 11,896 negative donations. Could be refunds, or redesignation. Yes they match

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [14]:
# 1k YOUR CODE HERE
len(contrib[(contrib['contb_receipt_dt'] < '2014-01-01') | (contrib['contb_receipt_dt'] > '2016-06-07')])

451372

In [15]:
contrib[(contrib['contb_receipt_dt'] < '2014-01-01') | (contrib['contb_receipt_dt'] > '2016-06-07')]['contb_receipt_dt'].head()

9932    2013-11-05
9994    2013-11-05
14673   2016-06-30
14682   2016-07-20
14697   2016-07-14
Name: contb_receipt_dt, dtype: datetime64[ns]

- **1k answer here:** Yes, there are 451,372 donations outside the primary period. Yes, the dates well formatted

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:** No, for rows, we need to filter for only the 2016 primary since there is 313,746 G2016 data. For columns, no, there are columns we dont need

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:** Yes and no. The zipcode column does not seem to be a zipcode but can be found using the city column to get an aproximate. Day data is not available but can be derived from the date column

**1l.3** What columns are important?

- **1l.3 answer here:**  cand_nm, contb_receipt_dt, contb_receipt_amt, tran_id, contbr_city, contbr_zip, contbr_st, and election_tp

**1l.4** What columns can be dropped?

- **1l.4 answer here:** cmte_id, contbr_nm, contbr_employer, contbr_occupation, receipt_desc, memo_cd, memo_text, file_num, form_tp, cand_id

**1l.5** What are the data problems?

- **1l.5 answer here:** The duplicate transactions, zipcodes do not seem like zipcodes, and days needs to be derived

**1l.6** List any assumptions so far:

- **1l.6 answer here:** Primary Period Definition, Transaction ID Uniqueness, That negative amounts are refunds or re-distributed donations, These are all valid candidates, Might use city and state to find zip code

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [16]:
# 2a YOUR CODE HERE
primary_data = contrib[contrib['election_tp'] == 'P2016']
primary_data.head()

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

In [17]:
primary_data.shape

(810481, 18)

**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [18]:
# 2b YOUR CODE HERE
primary_bernie_data = primary_data[primary_data['cand_nm'] == 'Sanders, Bernard']
primary_bernie_data.head()

cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
3  C00577130  P60007168  Sanders, Bernard         LEE, ALAN      CAMARILLO   
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   

  contbr_st     contbr_zip            contbr_employer  contbr_occupation  \
3        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER   
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION    PROJECT MANAGER   
8        CA 926,833,846.00                       NONE       NOT EMPLOYED   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
3              40.00       2016-03-04          NaN     NaN   
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   

  election_tp  
3       P2016  
4       P2016  
5       P2016  
6       P2016  
8       P2016

In [19]:
primary_bernie_data.shape

(407171, 18)

**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [20]:
# 2c YOUR CODE HERE
california_zipcodes_df = pd.read_csv('california_zipcodes.csv')
california_zipcodes_df.head()

OBJECTID  Zip Code         Post Office Name State  Pop 2017  \
0         1        12          Mt Meadows Area    CA       -99   
1         2        16  Sequoia National Forest    CA       -99   
2         3        17  Northeast Fresno County    CA       -99   
3         4        18    Los Padres Ntl Forest    CA       -99   
4         5        19        Lassen Ntl Forest    CA       -99   

   Pop SqMile 2017  SqMile       Shape__Area  Shape__Length  
0           -99.00   30.92    862,157,443.20     195,388.62  
1           -99.00   39.33  1,096,295,678.00     169,790.57  
2           -99.00  564.38 15,734,145,628.00     873,109.21  
3           -99.00   90.83  2,532,221,636.00     294,311.33  
4           -99.00  412.59 11,502,609,570.00     533,289.13

In [21]:
california_zipcodes_df.dtypes

OBJECTID              int64
Zip Code              int64
Post Office Name     object
State                object
Pop 2017              int64
Pop SqMile 2017     float64
SqMile              float64
Shape__Area         float64
Shape__Length       float64
dtype: object

In [22]:
primary_bernie_data['five_digit_zip'] = primary_bernie_data['contbr_zip'].astype(str).str.replace(r'\D', '').str[:5].astype(int)
primary_bernie_data.head()

C:\Users\mzkan\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
3  C00577130  P60007168  Sanders, Bernard         LEE, ALAN      CAMARILLO   
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   

  contbr_st     contbr_zip            contbr_employer  contbr_occupation  \
3        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER   
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION    PROJECT MANAGER   
8        CA 926,833,846.00                       NONE       NOT EMPLOYED   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
3              40.00       2016-03-04          NaN     NaN   
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   

  election_tp  five_digit_zip  
3       P2016           93011  
4       P2016           90278  
5       P2016           90278  
6       P2016           92084  
8       P2016           92683

In [23]:
primary_bernie_data['is_in_california'] = primary_bernie_data['five_digit_zip'].isin(california_zipcodes_df['Zip Code'])
primary_bernie_data.head()

C:\Users\mzkan\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
3  C00577130  P60007168  Sanders, Bernard         LEE, ALAN      CAMARILLO   
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   

  contbr_st     contbr_zip            contbr_employer  contbr_occupation  \
3        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER   
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION    PROJECT MANAGER   
8        CA 926,833,846.00                       NONE       NOT EMPLOYED   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
3              40.00       2016-03-04          NaN     NaN   
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   

  election_tp  five_digit_zip  is_in_california  
3       P2016           93011             False  
4       P2016           90278              True  
5       P2016           90278              True  
6       P2016           92084              True  
8       P2016           92683              True

In [24]:
primary_bernie_data = primary_bernie_data[primary_bernie_data['is_in_california'] == True]
primary_bernie_data.head()

cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   
9  C00577130  P60007168  Sanders, Bernard    LERCH, DOUGLAS       PETALUMA   

  contbr_st     contbr_zip            contbr_employer       contbr_occupation  \
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.              PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.              PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION         PROJECT MANAGER   
8        CA 926,833,846.00                       NONE            NOT EMPLOYED   
9        CA 949,522,729.00         SEEDS OF AWARENESS  DIRECTOR OF NON PROFIT   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   
9              15.00       2016-03-06          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   
9  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYY720   

  election_tp  five_digit_zip  is_in_california  
4       P2016           90278              True  
5       P2016           90278              True  
6       P2016           92084              True  
8       P2016           92683              True  
9       P2016           94952              True

In [25]:
primary_bernie_data.shape

(397509, 20)

**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [26]:
# 2d YOUR CODE HERE
primary_bernie_data = primary_bernie_data[primary_bernie_data['contb_receipt_amt'] >= 0]
primary_bernie_data.head()

cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   
9  C00577130  P60007168  Sanders, Bernard    LERCH, DOUGLAS       PETALUMA   

  contbr_st     contbr_zip            contbr_employer       contbr_occupation  \
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.              PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.              PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION         PROJECT MANAGER   
8        CA 926,833,846.00                       NONE            NOT EMPLOYED   
9        CA 949,522,729.00         SEEDS OF AWARENESS  DIRECTOR OF NON PROFIT   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   
9              15.00       2016-03-06          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   
9  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYY720   

  election_tp  five_digit_zip  is_in_california  
4       P2016           90278              True  
5       P2016           90278              True  
6       P2016           92084              True  
8       P2016           92683              True  
9       P2016           94952              True

In [27]:
primary_bernie_data.shape

(394484, 20)

**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [28]:
# 2e YOUR CODE HERE
columns_to_drop = ['cmte_id', 'cand_id', 'contbr_nm', 'contbr_city', 'contbr_st', 'contbr_zip', 'contbr_employer', \
                  'contbr_occupation', 'receipt_desc', 'memo_cd', 'memo_text', 'form_tp', 'file_num', \
                   'tran_id']

primary_bernie_data = primary_bernie_data.drop(columns=columns_to_drop)
primary_bernie_data.head()

cand_nm  contb_receipt_amt contb_receipt_dt election_tp  \
4  Sanders, Bernard              35.00       2016-03-05       P2016   
5  Sanders, Bernard             100.00       2016-03-06       P2016   
6  Sanders, Bernard              25.00       2016-03-04       P2016   
8  Sanders, Bernard              10.00       2016-03-05       P2016   
9  Sanders, Bernard              15.00       2016-03-06       P2016   

   five_digit_zip  is_in_california  
4           90278              True  
5           90278              True  
6           92084              True  
8           92683              True  
9           94952              True

In [29]:
primary_bernie_data.shape

(394484, 6)

- **2e answer here:** I only kept primary 2016 contribution data (election_tp) and concentrate on Bernie Sanders as a candidate (cand_nm), the amount, zipcode, is_in_california and date to answer the two questions from the boss

**2f.** List any assumptions that you made up to this point:

- **2f answer here:** California zip code file contains accurate and up-to-date information about valid zip codes in California, it was assumed that removing negative donation amounts from the analysis would improve the accuracy of the donation count and amount

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [30]:
# 3a YOUR CODE HERE
zipcode_stats = primary_bernie_data.groupby('five_digit_zip').agg({'contb_receipt_amt': ['count', 'sum']})
zipcode_stats.head()

contb_receipt_amt          
                           count       sum
five_digit_zip                            
90002                          3    545.00
90003                         15    180.00
90004                        744 40,785.35
90005                        174  7,985.76
90006                        247 10,509.76

In [31]:
zipcode_highest_count = zipcode_stats['contb_receipt_amt']['count'].idxmax()
zipcode_highest_count

94110

In [32]:
zipcode_highest_amount = zipcode_stats['contb_receipt_amt']['sum'].idxmax()
zipcode_highest_amount

94110

- **3a answer here:** 94110 had the highest contribution and count

**3b.** What day(s) of the month do most people donate? (10 points)

In [33]:
# 3b YOUR CODE HERE
primary_bernie_data['day'] = primary_bernie_data['contb_receipt_dt'].dt.day
day_stats = primary_bernie_data.groupby('day').agg({'contb_receipt_amt': ['count']})
day_stats.head()

contb_receipt_amt
                count
day                  
1               11613
2               10822
3                9268
4               11702
5               10287

In [34]:
most_donated_day = day_stats['contb_receipt_amt']['count'].idxmax()
most_donated_day

29

- **3b answer here:** The 29th day of the month is the most day that yhe number people donate 

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k